src/lib.rs

```diff
@@ -6,97 +9,107 @@ pub use websocket::AdminWebsocket;
 
 use std::fs;
 use std::io;
-use std::path::{Path, PathBuf};
+use std::path::Path;
 
 use anyhow::{anyhow, Context, Result};
-use futures::future::try_join_all;
-use tracing::{debug, info, instrument};
+use futures::{future::try_join_all, prelude::*};
+use tempfile::NamedTempFile;
+use tracing::{debug, info, instrument, trace};
 use url::Url;
 use zip::ZipArchive;
 
 #[instrument(err, fields(path = %path.as_ref().display()))]
 pub fn load_happs_yaml(path: impl AsRef<Path>) -> Result<Vec<Happ>> {
     use std::fs::File;
 
-    let file = File::open(path)?;
-    let happ_list = serde_yaml::from_reader(&file).context("failed to read happ list")?;
+    let file = File::open(path).context("failed to open file")?;
+    let happ_list =
+        serde_yaml::from_reader(&file).context("failed to deserialize YAML as Vec<Happ>")?;
     debug!(?happ_list);
     Ok(happ_list)
 }
 
 pub async fn install_happs(happ_list: &[Happ], config: &Config) -> Result<()> {
-    let mut admin_websocket = AdminWebsocket::connect(config.admin_port).await?;
-    let agent_key = admin_websocket.generate_agent_pubkey().await?;
-    debug!(?agent_key);
-
+    let admin_websocket = AdminWebsocket::connect(config.admin_port)
+        .await
+        .context("failed to connect to holochain")?;
     let futures: Vec<_> = happ_list
         .iter()
         .map(|happ| {
             let mut admin_websocket = admin_websocket.clone();
-            let agent_key = agent_key.clone();
-            let future = async move {
-                let install_happ = admin_websocket.install_happ(happ, agent_key, config.happ_port);
+            async move {
+                let mut agent_websocket = admin_websocket.clone();
```

Comment:
The dance around cloning `admin_websocket` again to generate `agent_key` is related to Rust borrowing rules. Without the second `.clone()`, `admin_websocket` is moved into the `and_then` closure and can't be used in the preceding future.